store/postgres/src/sql/parser.rs

```diff
@@ -0,0 +1,191 @@
+use super::{constants::SQL_DIALECT, formatter::Formatter, validation::Validator};
+use crate::relational::{ColumnType, Layout};
+use anyhow::{anyhow, Ok, Result};
+use graph::components::store::BLOCK_NUMBER_MAX;
+use itertools::Itertools;
+use std::sync::Arc;
+
+pub fn generate_table_prelude_from_layout(layout: &Layout) -> String {
+    let schema = &layout.catalog.site.namespace;
+    let ctes = layout
+        .tables
+        .iter()
+        .filter(|(entity, _)| !entity.is_poi())
+        .map(|(_, table)| {
+            let table_name = table.name.as_str();
+
+            let (block_column, filter) = if !table.immutable {
+                (
+                    "block_range",
+                    Some(format!(" WHERE \"block_range\" @> {}", BLOCK_NUMBER_MAX)),
+                )
+            } else {
+                ("block$", None)
+            };
+
+            let columns = table
+                .columns
+                .iter()
+                .map(|col| {
+                    if !col.is_list() && col.column_type == ColumnType::Bytes {
+                        format!(
+                            r#"concat('0x', encode("{}", 'hex')) AS "{}""#,
+                            col.name.as_str(),
+                            col.name.as_str()
+                        )
```

Comment:
Within internal tests, users struggle to understand that they can not use the `0x` prefix and tend to think that the data isn't there. Also, users believed that the UX was better when they received the prefix because they could ensure that it was an address.

It's easy to rewrite every string that starts with `0x` with `\x`, but if I'm not mistaken graph-node the responses from graphql queries are formatted somehow, we probably need to use it as well.

Reply:
You can't rewrite every `0x` string because quite a few subgraphs store hex data as strings and may or may not use a `0x` prefix for that - for strings, we don't know that as it's completely under the control of the mappings.

But this particular rewrite is an absolute performance killer. It will even rewrite expressions like `a.id = b.some_col`; that makes it impossible for Postgres to use the BTree index on those two columns, and might cause a full table scan depending in what other clauses the query contains.

For the UX problem, I see two options:
* document better that hex constants need to be written as `\xdeadbeef`
* do a bit of type inference on the AST; you'd want to infer for an expression like `a.col = '0x..'` that the literal needs to be a `bytea` based on the type of `a.col` and rewrite the literal in those cases. It's a little delicate since if you don't catch all cases (e.g., you also need to cover stuff like `a.col in ('0x..', ..)` and a lot of other expressions) it will confuse users even more since `0x..` will work in some places but not in others

There's another nuance of how indexes and queries interact: for most string and bytea columns, we have an index on a prefix of the value, not the whole value since that can be too big to put into an index. For GraphQL queries, the machinery that figures out whether to use a prefix or the whole column starts [here](https://github.com/graphprotocol/graph-node/blob/master/store/postgres/src/relational_queries.rs#L1464-L1473) If the query is not transformed to take advantage of those indexes, it again might become unbearably slow.